### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by R00BEL**

### CODE REVIEW

File: src/constants/colors.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const COLORS = {
```
Comment (pkarpovich) [inline src/constants/colors.js]: use CSS variable

File: src/constants/local-storage-keys.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const LocalStorageKeys = {
+  TOKEN: "token",
```
Comment (pkarpovich) [inline src/constants/local-storage-keys.js]: JWT_TOKEN

File: src/pages/back-office/sing-in/index.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+const SingIn = () => {
+  const [, setLocation] = useLocation();
+
+  const [error, setError] = useState(false);
+
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
+
+  const handleClick = useCallback(() => {
+    fetch("/api/users/singIn", {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: where your checks for password and correct email?

File: src/pages/back-office/sing-in/index.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      .then((res) => {
+        res.status !== 200 && setError(true);
+        return res.json();
+      })
+      .then((data) => {
+        localStorage.setItem(LocalStorageKeys.TOKEN, data);
+      });
+  }, [email, password]);
+
+  const transition = useCallback(() => {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: handleSignInButtonClick

File: src/pages/back-office/sing-in/index.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            type="password"
+            label="Password:"
+            name="password"
+            value={password}
+            onChange={setPassword}
+          />
+        </Flex>
+      </Flex>
+      <Flex direction="column" alignItems="center">
+        <Button onClick={handleClick}>Send</Button>
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: sing in

File: src/pages/back-office/sing-in/index.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      <Flex direction="column" mt={theme.spacing(1)}>
+        <Button onClick={transition}>Sing up</Button>
+      </Flex>
+    </s.Container>
+  );
+};
+
+const s = {
+  Container: styled(Flex)`
+    overflow-y: auto;
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: why U need this container?

File: src/pages/back-office/sing-in/index.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  `,
+  error: styled.p`
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: extract this error styles into shared component

File: src/pages/back-office/sing-up/index.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
+  const [repPassword, setRepPassword] = useState("");
+  const [checkEmail, setCheckEmail] = useState(false);
+  const [checkPassword, setChecktPassword] = useState(false);
+  const [checkRepPassword, setCheckRepPassword] = useState(false);
+
+  const handleClick = useCallback(() => {
+    const emailRe = /.@.../;
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: wrong regex, google more

File: src/pages/back-office/sing-up/index.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    if (emailRe.test(email) && email.length < 4) {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: all these checks should be in regex

File: src/pages/back-office/sing-up/index.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      </Flex>
+
+      <Flex direction="column" mt={theme.spacing(1)}>
+        <Button onClick={transition}>Sing in</Button>
+      </Flex>
+    </s.Container>
+  );
+};
+
+const s = {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: same as above

**Review by pkarpovich — CHANGES_REQUESTED** [review]

File: src/pages/back-office/sing-in/index.js [hunk 11]
```diff
@@ -1,96 +1,128 @@
...
-      .then((res) => {
-        res.status !== 200 && setError(true);
-        return res.json();
+  const formik = useFormik({
+    initialValues: {
+      email: "",
+      password: "",
+    },
+    validationSchema: SignupSchema,
+    onSubmit: (values) => {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: use callback

File: src/pages/back-office/sing-in/index.js [hunk 12]
```diff
@@ -1,96 +1,128 @@
...
   const [, setLocation] = useLocation();
+  const {
+    strings: { singIn: translations },
+  } = useTranslation();
 
   const [error, setError] = useState(false);
 
-  const [email, setEmail] = useState("");
-  const [password, setPassword] = useState("");
+  const SignupSchema = Yup.object().shape({
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js:37]: you recreate schema for every render. Extract from the component body

File: src/pages/back-office/sing-in/index.js [hunk 13]
```diff
@@ -1,96 +1,128 @@
...
-      body: JSON.stringify({
-        email: email,
-        password: password,
-      }),
-    })
-      .then((res) => {
-        res.status !== 200 && setError(true);
-        return res.json();
+  const formik = useFormik({
+    initialValues: {
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: extract from component into const

File: src/pages/back-office/sing-in/index.js [hunk 14]
```diff
@@ -1,96 +1,128 @@
...
+            <Flex direction="column">
+              <InputFormik
+                id="password"
+                type="password"
+                label={translations["password"]}
+                value={formik.values.password}
+                onChange={formik.handleChange}
+              />
+            </Flex>
+            {formik.touched.password && formik.errors.password ? (
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js:127]: why you need touched in this case? errors will enough

File: src/pages/back-office/sing-in/index.js [hunk 15]
```diff
@@ -1,96 +1,128 @@
...
+              <Error>{formik.errors.password}</Error>
+            ) : null}
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js:129]: just &&, without ? :

File: src/pages/back-office/sing-in/index.js [hunk 16]
```diff
@@ -1,96 +1,128 @@
...
-  Container: styled(Flex)`
-    overflow-y: auto;
-  `,
-  error: styled.p`
-    color: ${COLORS.ERROR};
-    font-size: 12px;
-    margin: 0;
-  `,
-};
+const Error = styled.p`
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: error text and separate file

File: src/pages/back-office/sing-up/index.js [hunk 17]
```diff
@@ -1,140 +1,143 @@
...
-    } else {
-      setCheckRepPassword(false);
-    }
-
-    if (
-      emailRe.test(email) &&
-      password.length >= 8 &&
-      password === repPassword
-    ) {
+  const SignupSchema = Yup.object().shape({
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: same as below

File: src/pages/back-office/sing-up/index.js [hunk 18]
```diff
@@ -1,140 +1,143 @@
...
+            <Flex direction="column">
+              <InputFormik
+                id="email"
+                type="string"
+                label={translations["email"]}
+                value={formik.values.email}
+                onChange={formik.handleChange}
+              />
+            </Flex>
+            {formik.touched.email && formik.errors.email ? (
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: same as below

File: src/components/input-formik.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+InputFormik.propTypes = {
+  id: PropTypes.string,
+  type: PropTypes.string,
+  label: PropTypes.string,
+  name: PropTypes.string,
+  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  onChange: PropTypes.func,
+};
+
+const StyledInput = styled.input`
```
Comment (pkarpovich) [inline src/components/input-formik.js]: s object, use this style

File: src/components/input-formik.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
+import { memo } from "react";
+import PropTypes from "prop-types";
+import { styled } from "@linaria/react";
+import { Label } from "./label";
+
+const InputFormik = ({ type = "text", label, name, value, onChange, id }) => {
```
Comment (pkarpovich) [inline src/components/input-formik.js]: we already have an input component, why U create new

**Review by pkarpovich — CHANGES_REQUESTED** [review]

File: src/pages/back-office/sing-in/index.js (hunk 13 above)
Comment (R00BEL) [reply src/pages/back-office/sing-in/index.js]: corrected

File: src/components/input-formik.js (hunk 19 above)
Comment (R00BEL) [reply src/components/input-formik.js]: input-formik was removed

File: src/components/input-formik.js (hunk 20 above)
Comment (R00BEL) [reply src/components/input-formik.js]: fixed

File: src/App.js [hunk 21]
```diff
@@ -38,24 +42,29 @@ function App() {
             <ErrorBoundary>
               <Switch>
                 <Route
-                    path={Routes.SING_UP}
-                    component={SingUp}
-                    layout={MainLayout}
+                  path={Routes.PRIVATE_ROUTE}
```
Comment (pkarpovich) [inline src/App.js]: Add route for public and private router component, in which you will check JWT token

File: src/constants/routes.js [hunk 22]
```diff
@@ -5,4 +5,5 @@ export const Routes = {
   QR_CODE_BUILDER: "/qr-code-builder",
   SING_UP: "/back-office/sing-up",
   SING_IN: "/back-office/sing-in",
+  PRIVATE_ROUTE: "/private-route",
```
Comment (pkarpovich) [inline src/constants/routes.js]: /back-office/dashboard

File: src/pages/back-office/sing-up/index.js [hunk 23]
```diff
@@ -75,53 +81,55 @@ const SingUp = () => {
...
             <Flex direction="column">
-              <InputFormik
+              <Input
                 id="email"
                 type="string"
                 label={translations["email"]}
-                value={formik.values.email}
-                onChange={formik.handleChange}
+                value={formik.values["email"]}
+                onChange={(e) => formik.setFieldValue("email", e)}
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: useCallback

File: src/pages/back-office/sing-up/index.js [hunk 24]
```diff
@@ -75,53 +81,55 @@ const SingUp = () => {
         <Flex direction="column">
           <Flex mb={theme.spacing(1)} direction="column">
             <Flex direction="column">
-              <InputFormik
+              <Input
                 id="email"
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js:96]: why you need id?

File: src/pages/back-office/sing-up/index.js [hunk 25]
```diff
@@ -75,53 +81,55 @@ const SingUp = () => {
...
+            {formik.touched.email && formik.errors.email && (
+              <ErrorText>{formik.errors.email}</ErrorText>
+            )}
           </Flex>
 
           <Flex mb={theme.spacing(1)} direction="column">
             <Flex direction="column">
-              <InputFormik
+              <Input
                 id="password"
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js:111]: same as below

File: src/pages/back-office/sing-up/index.js [hunk 26]
```diff
@@ -75,53 +81,55 @@ const SingUp = () => {
...
                 type="password"
                 label={translations["password"]}
-                value={formik.values.password}
-                onChange={formik.handleChange}
+                value={formik.values["password"]}
+                onChange={(e) => formik.setFieldValue("password", e)}
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: same as below

File: src/pages/back-office/sing-up/index.js [hunk 27]
```diff
@@ -75,53 +81,55 @@ const SingUp = () => {
...
             <Flex direction="column">
-              <InputFormik
+              <Input
                 id="repPassword"
                 type="password"
                 label={translations["repeat_password"]}
-                value={formik.values.repPassword}
-                onChange={formik.handleChange}
+                value={formik.values["repPassword"]}
+                onChange={(e) => formik.setFieldValue("repPassword", e)}
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: same as below

**Review by pkarpovich — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 21 above)
Comment (R00BEL) [reply src/App.js]: added

File: src/constants/routes.js (hunk 22 above)
Comment (R00BEL) [reply src/constants/routes.js]: fix

File: src/pages/back-office/sing-up/index.js (hunk 23 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: fix

File: src/pages/back-office/sing-up/index.js (hunk 26 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: fix

File: src/pages/back-office/sing-up/index.js (hunk 27 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: fix

File: src/App.js [hunk 28]
```diff
@@ -29,8 +30,8 @@ const SingIn = lazy(() =>
   import("pages/back-office/sing-in" /* webpackChunkName: "sing-in" */)
 );
 
-const PrivateRoute = lazy(() =>
-  import("pages/public/private-route" /* webpackChunkName: "private-route" */)
+const dashboard = lazy(() =>
```
Comment (pkarpovich) [inline src/App.js]: Dashboard

File: src/components/private-route.js [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+import Route from "./Route";
+import { LocalStorageKeys } from "../constants/local-storage-keys";
+import { Routes } from "../constants/routes";
+import { useLocation } from "wouter";
+
+const PrivateRoute = (props) => {
+  const [, setLocation] = useLocation();
+
+  useEffect(() => {
+    const token = decodeToken(localStorage.getItem(LocalStorageKeys.JWT_TOKEN));
```
Comment (pkarpovich) [inline src/components/private-route.js]: isExpired

File: src/components/private-route.js [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+
+    if (!token) {
+      setLocation(Routes.SING_IN);
+    } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(token.email)) {
```
Comment (pkarpovich) [inline src/components/private-route.js]: why U need this?

File: src/pages/back-office/sing-in/index.js [hunk 31]
```diff
@@ -86,7 +87,9 @@ const SingIn = () => {
                 type="string"
                 label={translations["email"]}
                 value={formik.values["email"]}
-                onChange={(e) => formik.setFieldValue("email", e)}
+                onChange={useCallback((e) => formik.setFieldValue("email", e), [
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: Seriously? Why U not create a function? It looks unusually

File: src/pages/back-office/sing-in/index.js [hunk 32]
```diff
@@ -101,7 +104,10 @@ const SingIn = () => {
                 type="password"
                 label={translations["password"]}
                 value={formik.values["password"]}
-                onChange={(e) => formik.setFieldValue("password", e)}
+                onChange={useCallback(
```
Comment (pkarpovich) [inline src/pages/back-office/sing-in/index.js]: the same

File: src/pages/back-office/sing-up/index.js [hunk 33]
```diff
@@ -86,7 +86,9 @@ const SingUp = () => {
                 type="string"
                 label={translations["email"]}
                 value={formik.values["email"]}
-                onChange={(e) => formik.setFieldValue("email", e)}
+                onChange={useCallback((e) => formik.setFieldValue("email", e), [
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: the same

File: src/pages/back-office/sing-up/index.js [hunk 34]
```diff
@@ -101,7 +103,10 @@ const SingUp = () => {
                 type="password"
                 label={translations["password"]}
                 value={formik.values["password"]}
-                onChange={(e) => formik.setFieldValue("password", e)}
+                onChange={useCallback(
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: the same

File: src/pages/back-office/sing-up/index.js [hunk 35]
```diff
@@ -116,7 +121,10 @@ const SingUp = () => {
                 type="password"
                 label={translations["repeat_password"]}
                 value={formik.values["repPassword"]}
-                onChange={(e) => formik.setFieldValue("repPassword", e)}
+                onChange={useCallback(
```
Comment (pkarpovich) [inline src/pages/back-office/sing-up/index.js]: the same

**Review by pkarpovich — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 28 above)
Comment (R00BEL) [reply src/App.js]: fix

File: src/components/private-route.js (hunk 29 above)
Comment (R00BEL) [reply src/components/private-route.js]: fix

File: src/components/private-route.js (hunk 30 above)
Comment (R00BEL) [reply src/components/private-route.js]: fix

File: src/pages/back-office/sing-in/index.js (hunk 31 above)
Comment (R00BEL) [reply src/pages/back-office/sing-in/index.js]: added handleChange function

File: src/pages/back-office/sing-in/index.js (hunk 32 above)
Comment (R00BEL) [reply src/pages/back-office/sing-in/index.js]: added handleChange function

File: src/pages/back-office/sing-up/index.js (hunk 33 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: added handleChange function

File: src/pages/back-office/sing-up/index.js (hunk 34 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: added handleChange function

File: src/pages/back-office/sing-up/index.js (hunk 35 above)
Comment (R00BEL) [reply src/pages/back-office/sing-up/index.js]: added handleChange function

**Review by pkarpovich — APPROVED** [review]

**Pull request merged**